src/dbus/mutter_display_config.rs

```diff
@@ -314,11 +330,110 @@ impl DisplayConfig {
+                    .find(|o| o.name == connector)
+                    .is_none()
+                {
+                    return Err(zbus::fdo::Error::Failed(format!(
+                        "Connector '{}' not found",
+                        connector
+                    )));
+                }
+                let msgs = vec![
+                    OutputAction::Mode {
+                        mode: ModeToSet::Specific(ConfiguredMode::from_str(&mode).map_err(
+                            |e| {
+                                zbus::fdo::Error::Failed(format!(
+                                    "Could not parse mode '{}': {}",
+                                    mode, e
+                                ))
+                            },
+                        )?),
+                    },
+                    OutputAction::Position {
+                        position: PositionToSet::Specific(ConfiguredPosition {
+                            x: requested_config.x,
+                            y: requested_config.y,
+                        }),
+                    },
+                    OutputAction::Scale {
+                        scale: ScaleToSet::Specific(requested_config.scale),
+                    },
+                    OutputAction::Transform {
+                        transform: match requested_config.transform {
+                            0 => niri_ipc::Transform::Normal,
+                            1 => niri_ipc::Transform::_90,
+                            2 => niri_ipc::Transform::_180,
+                            3 => niri_ipc::Transform::_270,
+                            4 => niri_ipc::Transform::Flipped,
+                            5 => niri_ipc::Transform::Flipped90,
+                            6 => niri_ipc::Transform::Flipped180,
+                            7 => niri_ipc::Transform::Flipped270,
+                            x => {
+                                return Err(zbus::fdo::Error::Failed(format!(
+                                    "Unknown transform {}",
+                                    x
+                                )))
+                            }
+                        },
+                    },
+                    OutputAction::On,
+                ];
+                messages.insert(connector, msgs);
+            }
+        }
+        if messages.is_empty() {
+            return Err(zbus::fdo::Error::Failed(
+                "At least one output must be enabled".to_owned(),
+            ));
+        }
+        if method == 0 {
+            // 0 means "verify", so don't actually apply here
+            return Ok(());
+        }
+        for (_, output) in current_conf.iter() {
+            if let Some(msgs) = messages.remove(&output.name) {
+                for msg in msgs.into_iter() {
+                    self.send(&output.name, msg)?;
+                }
+            } else {
+                self.send(&output.name, OutputAction::Off)?;
+            }
+        }
```

Comment:
The config should be applied atomically, i.e. as one single message on the channel for all changes.